### PR TITLE
Implement generator as a Rails template

### PR DESCRIPTION
**Pull request by gsmendoza**
## Goal

As a SolidusStarterFrontend Contributor
I want to implement the SolidusStarterFrontend generator as a Rails template
So that it would be easier to install it on a fresh Rails application

## Direction

Based on my discussions with @kennyadsl, we've chosen to simplify the project, opting for a small, opinionated solution. As a result, we've removed several existing and planned features from the project, including:

* Option to skip specs when installing the frontend
* Option to skip the authentication components of the frontend
* Rails engine used in running the sandbox app in development
* Gem caching in CI

## Walkthrough

https://www.loom.com/share/1a3cd2d285e548edbe2eb3d8e389aeb9

## Demo

https://www.loom.com/share/1f3db0216d7045e599250922f55a5d4d

## How Has This Been Tested?

* [x] Tested remote install (see demo above)
* [x] Tested installing sandbox app by calling sandbox script directly
* [x] Tested installing sandbox app in a Docker container (see [development.md](https://github.com/nebulab/solidus_starter_frontend/blob/master/docs/development.md))

## Previous implementations

* https://github.com/nebulab/solidus_starter_frontend/pull/185
* https://github.com/nebulab/solidus_starter_frontend/pull/184
* https://github.com/nebulab/solidus_starter_frontend/pull/182

## Types of changes

- N/A: Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: template.rb [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+  Bundler.with_original_env do
+    run 'bundle install'
+  end
+end
+
+def add_solidus_gems
+  solidus_repo = ENV['SOLIDUS_REPO']
+  solidus_branch = ENV['SOLIDUS_BRANCH']
+
+  if solidus_repo && solidus_branch
```
Comment (gsmendoza) [inline template.rb:59]: @aldesantis FYI, I added this `solidus_repo` check after I made the walkthrough. Previously, the code only checked `solidus_branch` and set `solidus_repo` to the default only if `solidus_branch` is provided. You can see the previous code in the walkthrough.

File: bin/solidus_compare [hunk 2]
```diff
@@ -96,13 +96,14 @@ def load_config
   @config['source_name'] ||= 'solidus'
   @config['source_branch'] ||= 'master'
   @config['source_base_path'] ||= 'frontend/'
+  @config['local_base_path'] ||= 'templates/'
```
[the remark this answers is not in the document]
Comment (gsmendoza) [reply bin/solidus_compare:99]: @aldesantis Perhaps we can remove it. I think the only consequence is that we'd _still_ have to manually watch `solidus_frontend` for changes and remember to update this repo if that ever that happens. Considering that this tool doesn't need much maintenance or resources, I think it's smarter to keep it :)

File: template.rb [hunk 3]
```diff
@@ -33,6 +34,31 @@ def add_template_repository_to_source_path
...
+  gem 'solidus_core', github: solidus_repo, branch: solidus_branch
+  gem 'solidus_api', github: solidus_repo, branch: solidus_branch
+  gem 'solidus_backend', github: solidus_repo, branch: solidus_branch
+  gem 'solidus_sample', github: solidus_repo, branch: solidus_branch
+else
+  gem 'solidus_core'
+  gem 'solidus_api'
+  gem 'solidus_backend'
+  gem 'solidus_sample'
+end
```
[the remark this answers is not in the document]
Comment (gsmendoza) [reply template.rb]: > Instead, why don't we replace this with instructions in the readme about how to generate a new Solidus app without the default storefront and then install the new storefront?

@aldesantis The "Solidus with starter frontend" template was the solution that @kennyadsl and I decided to implement. However, my previous implementation of this feature is more in line with what you're suggesting. Please see the commits from "Copy add_template_repository_to_source_path method" (404f96463a69867b3be94176cbc19d2115802d95) to Remove generator along with gem (cb751649d5ddc292cc71822db9f7d156d6e0ff53) in this PR. In this previous implementation, the user has to generate a new Solidus app first before installing the storefront.

I've also experimented with adding a `with_frontend` option to `solidus:install` that can be used in conjuction with the frontend-only template. Please see https://github.com/solidusio/solidus/pull/4181/files.

While we're waiting for @kennyadsl, perhaps we can merge the frontend-only template first. Like you said, I can submit a separate PR later on for the "Solidus with starter frontend" template. 

What do you think?

File: templates/lib/solidus_starter_frontend/config.rb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+require 'solidus_starter_frontend_configuration'
```
[the remark this answers is not in the document]
Comment (gsmendoza) [reply templates/lib/solidus_starter_frontend/config.rb]: Thanks! I didn't notice that. I removed those in `app/lib`.

File: templates/lib/solidus_starter_frontend_configuration.rb [hunk 5]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+class SolidusStarterFrontendConfiguration < Spree::Preferences::Configuration
```
[the remark this answers is not in the document]
Comment (gsmendoza) [reply templates/lib/solidus_starter_frontend_configuration.rb]: Thanks! I didn't notice that. I removed those in `app/lib`.

File: template.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (aldesantis) [reply template.rb]: Yep, let's remove this for the time being and we can include it in a future PR if we agree it's a good idea.

**Review by aldesantis — APPROVED** [review]
This looks good, thanks @gsmendoza for the thorough work!

**Pull request merged**